### PR TITLE
Integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by muhammedb8826**
### In this Milestone we have worked on the following:

N + 1 Improvements

User#index before

user index before

User#index after

user index after

User#show before

user show before

User#show after

user show after

- User index page:

I can see the username of all other users.
I can see the profile picture for each user.
I can see the number of posts each user has written.
When I click on a user, I am redirected to that user's show page.

- user show page:

I can see the user's profile picture.
I can see the user's username.
I can see the number of posts the user has written.
I can see the user's bio.
I can see the user's first 3 posts.
I can see a button that lets me view all of a user's posts.
When I click a user's post, it redirects me to that post's show page.
When I click to see all posts, it redirects me to the user's post's index page.

- User post index page:

I can see the user's profile picture.
I can see the user's username.
I can see the number of posts the user has written.
I can see a post's title.
I can see some of the post's body.
I can see the first comments on a post.
I can see how many comments a post has.
I can see how many likes a post has.
I can see a section for pagination if there are more posts than fit on the view.
When I click on a post, it redirects me to that post's show page.

- Post show page:

I can see the post's title.
I can see who wrote the post.
I can see how many comments it has.
I can see how many likes it has.
I can see the post body.
I can see the username of each commentor.
I can see the comment each commentor left.

- All the tests passing

all the tests are passed

### CODE REVIEW

File: spec/integration/user_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+      visit "users/#{user_two.id}"
+      user_two.recent_posts.each do |post|
+        expect(page).to have_link(post.title, href: "/users/#{user_two.id}/posts/#{post.id}")
+      end
+    end
+    it 'should render a button that lets me view all of a users posts' do
+      visit "users/#{user_two.id}"
+      expect(page).to have_link('See all posts', href: "/users/#{user_two.id}/posts")
+    end
+  end
```
Comment (codepantha) [inline spec/integration/user_show_spec.rb:69]: - [ ] There are 2 tests shown below that are missing in your `user_show_spec.rb` file. Please write them as they compulsory requirements 👍 
image

File: spec/integration/posts_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  let!(:comment_two) { Comment.create author: user_two, post: post_two, text: 'hello' }
+  let!(:comment_three) { Comment.create author: user_two, post: post_two, text: 'hello' }
+  let!(:comment_four) { Comment.create author: user_two, post: post_two, text: 'hello' }
+  let!(:comment_five) { Comment.create author: user_two, post: post_two, text: 'hello' }
+  let!(:comment_six) { Comment.create author: user_two, post: post_two, text: 'hello' }
+  # I can see the user's username.
+  it 'should render the user name' do
+    visit "users/#{user_two.id}/posts"
+    expect(page).to have_content(user_two.name)
+  end
```
Comment (codepantha) [inline spec/integration/posts_index_spec.rb]: - [ ] Please write a test that asserts you can see the user's profile picture 👍

File: spec/integration/user_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+      expect(page).to have_content(user_two.name)
+    end
+    it 'should render the users photo' do
+      visit users_path
+      expect(page).to have_content(user_two.photo)
+    end
+    it 'should render the number of posts' do
+      visit users_path
+      expect(page).to have_content("Number of posts #{user_two.posts.size}")
+    end
```
Comment (codepantha) [inline spec/integration/user_index_spec.rb:17]: - [ ] In the testing requirements for the User index page, it is expected that you write tests that assert an expected outcome for all users - not a single user as you have done here. Please see the requirements below and adjust each of your test to cater to all the users 🙏 
image

**Review by codepantha — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED ♻️

## Hi team 👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights 🎯 
✔️ Descriptive PR
✔️ Written integration tests 📌 
✔️ No linter errors

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- n/a

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Reem-lab — APPROVED** [review]
# Hi @muhammedb8826,  👋🏻👋🏻

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


## Highlights

- [ ] No Linters errors 🔥
- [ ]  Great job working on the functionality of the tests💯
- [ ]  Followed Gitflow 🔥
- [ ] Kudos for working on the requested changes 💯🤩

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**